enterprise/coderd/coderd.go

```diff
@@ -2,48 +2,218 @@ package coderd
+		r.Route("/licenses", func(r chi.Router) {
+			r.Use(apiKeyMiddleware)
+			r.Post("/", api.postLicense)
+			r.Get("/", api.licenses)
+			r.Delete("/{id}", api.deleteLicense)
+		})
+	})
+
+	err := api.updateEntitlements(ctx)
+	if err != nil {
+		return nil, xerrors.Errorf("update entitlements: %w", err)
+	}
+	api.closeLicenseSubscribe, err = api.Pubsub.Subscribe(pubSubEventLicenses, func(ctx context.Context, message []byte) {
+		_ = api.updateEntitlements(ctx)
+	})
+	if err != nil {
+		return nil, xerrors.Errorf("subscribe to license updates: %w", err)
+	}
+	go api.runEntitlementsLoop(ctx)
+
+	return api, nil
+}
+
+type Options struct {
+	*coderd.Options
+
+	EntitlementsUpdateInterval time.Duration
+	Keys                       map[string]ed25519.PublicKey
+}
+
+type API struct {
+	AGPL *coderd.API
+	*Options
+
+	closeLicenseSubscribe  func()
+	cancelEntitlementsLoop func()
+	mutex                  sync.RWMutex
+	hasLicense             bool
+	activeUsers            codersdk.Feature
+	auditLogs              codersdk.Entitlement
+}
+
+func (api *API) Close() error {
+	api.closeLicenseSubscribe()
+	api.cancelEntitlementsLoop()
+	return api.AGPL.Close()
+}
+
+func (api *API) runEntitlementsLoop(ctx context.Context) {
+	ticker := time.NewTicker(api.EntitlementsUpdateInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+		}
+		err := api.updateEntitlements(ctx)
+		if err != nil {
+			api.Logger.Warn(ctx, "failed to get feature entitlements", slog.Error(err))
+			continue
+		}
+	}
+}
+
+func (api *API) updateEntitlements(ctx context.Context) error {
+	licenses, err := api.Database.GetUnexpiredLicenses(ctx)
+	if err != nil {
+		return err
+	}
+	api.mutex.Lock()
+	defer api.mutex.Unlock()
+	now := time.Now()
+	auditLogs := api.auditLogs
+	for _, l := range licenses {
+		claims, err := validateDBLicense(l, api.Keys)
+		if err != nil {
+			api.Logger.Debug(ctx, "skipping invalid license",
+				slog.F("id", l.ID), slog.Error(err))
+			continue
+		}
+		api.hasLicense = true
+		entitlement := codersdk.EntitlementEntitled
+		if now.After(claims.LicenseExpires.Time) {
+			// if the grace period were over, the validation fails, so if we are after
+			// LicenseExpires we must be in grace period.
+			entitlement = codersdk.EntitlementGracePeriod
+		}
+		if claims.Features.UserLimit > 0 {
+			api.activeUsers.Enabled = true
+			api.activeUsers.Entitlement = entitlement
+			currentLimit := int64(0)
+			if api.activeUsers.Limit != nil {
+				currentLimit = *api.activeUsers.Limit
+			}
+			limit := max(currentLimit, claims.Features.UserLimit)
+			api.activeUsers.Limit = &limit
+		}
+		if claims.Features.AuditLog > 0 {
+			api.auditLogs = entitlement
+		}
+	}
+	if auditLogs != api.auditLogs {
+		auditor := agplaudit.NewNop()
+		// A flag could be added to the options that would allow disabling
+		// enhanced audit logging here!
+		if api.auditLogs == codersdk.EntitlementEntitled {
```

Comment:
This changes audit logging to move into hard enforcement with no grace-period, which we explicitly said will piss customers off.

If we allow grace periods, we must allow disabling audit logs, since we will start warning customers that they are in violation of their license, so they must have an action they can take (short of removing Coder entirely) to get back into compliance.

Reply:
Good point, I'll add that back in.